File: frontend/runtime_console_v3/src/lib/tcg-adapters/swu.ts
import { parseGenericDecklist } from "@/lib/tcg-adapters/parse-decklist";
import type { GameAdapter, TournamentFormat } from "@/lib/tcg-adapters/types";

const FORMATS: TournamentFormat[] = [
  { code: "STANDARD", name: "Standard", decklist_required: true, decklist_validation: true, default_timer_minutes: 55, default_match_type: "BO3", default_swiss_rounds: "auto", default_top_cut: 8 },
  { code: "LIMITED", name: "Limited", decklist_required: false, decklist_validation: false, default_timer_minutes: 55, default_match_type: "BO3", default_swiss_rounds: "auto", default_top_cut: 8 },
];

export const swuAdapter: GameAdapter = {
  code: "SWU",
  name: "Star Wars Unlimited",
  slug: "swu",
  supportedFormats: FORMATS,
  parseDecklist: (raw, format) => parseGenericDecklist(raw, "swu", format),
  getDecklistHint: () => "Exatamente 50 cartas. Máx. 3 cópias por carta.",
  getMaxCopies: () => 3,
  getMinDeckSize: () => 50,
  getMaxDeckSize: () => 50,
  getSideboardRules: () => null,
};

File: services/api/app/tcg_adapters/swu.py
"""Adaptador Star Wars Unlimited (referência existente)."""

from __future__ import annotations

from app.tcg_adapters.base import GameAdapter
from app.tcg_adapters.parsers import parse_generic_lines
from app.tcg_adapters.types import (
    ParsedDecklist,
    TournamentFormatConfig,
    UnifiedCard,
    ValidationResult,
    normalize_card_name,
)
from app.tcg_adapters.validation import (
    build_result,
    validate_copies,
    validate_size,
)


class SwuAdapter(GameAdapter):
    code = "SWU"  # type: ignore[assignment]
    name = "Star Wars Unlimited"

    _SEED_CARDS = [
        UnifiedCard(
            id="luke-skywalker",
            game="SWU",
            external_id="luke-skywalker",
            name="Luke Skywalker",
            normalized_name="luke skywalker",
            set_code="SOR",
            card_type="Unit",
            game_specific_type="leader",
            legality={"STANDARD": "legal"},
            game_data={"aspect": "Heroism", "cost": 5},
        ),
    ]

    def supported_formats(self) -> list[TournamentFormatConfig]:
        return [
            TournamentFormatConfig(
                "STANDARD", "Standard", "Constructed SWU",
                True, True, 55, "BO3", "auto", 8, 4, 512,
            ),
            TournamentFormatConfig(
                "LIMITED", "Limited", "Draft / Sealed SWU",
                False, False, 55, "BO3", "auto", 8, 4, 512,
            ),
        ]

    def parse_decklist(self, raw: str, fmt: str) -> ParsedDecklist:
        return parse_generic_lines(raw, "swu", fmt)

    def validate_decklist(self, deck: ParsedDecklist) -> ValidationResult:
        fmt = deck.format.upper()
        errors = []
        if fmt == "LIMITED":
            return build_result(deck, errors)

        errors.extend(validate_size(deck, min_size=50, max_size=50, exact=True))
        errors.extend(validate_copies(deck, max_copies=3, use_normalized=True))
        return build_result(deck, errors)

    def get_max_copies(self, fmt: str) -> int:
        return 3

    def get_min_deck_size(self, fmt: str) -> int:
        return 50

    def get_max_deck_size(self, fmt: str) -> int | None:
        return 50

    def search_cards(self, query: str, limit: int = 20) -> list[UnifiedCard]:
        q = normalize_card_name(query)
        return [c for c in self._SEED_CARDS if q in c.normalized_name][:limit]
